vendor/k8s.io/gengo/v2/comments.go

```diff
@@ -81,3 +83,202 @@ func ExtractSingleBoolCommentTag(marker string, key string, defaultVal bool, lin
+// may match lines which fail this syntax (e.g. which predate this definition).
+//
+// This function looks for input lines of the following forms:
+//   - 'marker' + "key=value"
+//   - 'marker' + "key()=value"
+//   - 'marker' + "key(arg)=value"
+//
+// The arg is optional.  If not specified (either as "key=value" or as
+// "key()=value"), the resulting Tag will have an empty Args list.
+//
+// The value is optional.  If not specified, the resulting Tag will have "" as
+// the value.
+//
+// Tag comment-lines may have a trailing end-of-line comment.
+//
+// The map returned here is keyed by the Tag's name without args.
+//
+// A tag can be specified more than one time and all values are returned.  If
+// the resulting map has an entry for a key, the value (a slice) is guaranteed
+// to have at least 1 element.
+//
+// Example: if you pass "+" as the marker, and the following lines are in
+// the comments:
+//
+//	+foo=val1  // foo
+//	+bar
+//	+foo=val2  // also foo
+//	+baz="qux"
+//	+foo(arg)  // still foo
+//
+// Then this function will return:
+//
+//		map[string][]Tag{
+//	 	"foo": []Tag{{
+//				Name: "foo",
+//				Args: nil,
+//				Value: "val1",
+//			}, {
+//				Name: "foo",
+//				Args: nil,
+//				Value: "val2",
+//			}, {
+//				Name: "foo",
+//				Args: []string{"arg"},
+//				Value: "",
+//			}, {
+//				Name: "bar",
+//				Args: nil,
+//				Value: ""
+//			}, {
+//				Name: "baz",
+//				Args: nil,
+//				Value: "\"qux\""
+//		}}
+func ExtractFunctionStyleCommentTags(marker string, prefixes []string, lines []string) (map[string][]Tag, error) {
+	stripComment := func(in string) string {
+		parts := strings.SplitN(in, "//", 2)
+		return strings.TrimSpace(parts[0])
+	}
+
+	matches := func(line string) bool {
+		if len(line) == 0 {
+			return false
+		}
+		if !strings.HasPrefix(line, marker) {
+			return false
+		}
+		if len(prefixes) == 0 {
+			//FIXME: should this fail, forcing callers to pass prefixes?  Not
+			// passing them fails when it hits real tags like:
+			// +protobuf.options.(gogoproto.goproto_stringer)=false
+			return true
+		}
+		for _, pfx := range prefixes {
+			if strings.HasPrefix(line, marker+pfx) {
+				return true
+			}
+		}
+		return false
+	}
+
+	out := map[string][]Tag{}
+	for _, line := range lines {
+		line = strings.TrimSpace(line)
+		if !matches(line) {
+			continue
+		}
+		line = stripComment(line)
+		kv := strings.SplitN(line[len(marker):], "=", 2)
+		key := kv[0]
+		val := ""
+		if len(kv) == 2 {
+			val = kv[1]
+		}
+
+		tag := Tag{}
+		if name, args, err := parseTagKey(key); err != nil {
+			return nil, err
+		} else {
+			tag.Name, tag.Args = name, args
+		}
+		tag.Value = val
+
+		out[tag.Name] = append(out[tag.Name], tag)
+	}
+	return out, nil
+}
+
+// Tag represents a single comment tag.
+type Tag struct {
+	// Name is the name of the tag with no arguments.
+	Name string
+	// Args is a list of optional arguments to the tag.
+	Args []string
```

Comment:
IIUC, "Args" here adds one more dimension for the `Tag`. How about just treat it as a string and let the caller to handle it?

Reply:
Aaron did `subfield` that way, and it works because the search space is small.  Given a struct with 3 fields (A, B, C), there are exactly 3 tag names to search for: "subfield(A)", "subfield(B)", "subfield(C)".

There's an unbounded number of options to search for, so running the Extract function for each one is sort of painful.